### PR TITLE
feature(stage): watch file changes in repository when window is inactive

### PR DESCRIPTION
**Pull request by calahatras**
Refresh status and selected file if any change occurs when window is inactive

Solves go-27

### CODE REVIEW

**Review by mickedefault — COMMENTED** [review]
Code looks great! Haven't tested it (I'm on my phone).
I'm wondering if we should make this togglable or present a button to explicitly refresh 🤔 sometimes you might *want* to see how it looked when you last had go open (something like "crap i made a lot of changes here.. and i didn't do a commit like i should have. Go, please show me how it looked before i changed this")

**Comment by calahatras** [comment]
Yeah, had that first and then just figured this was easier. 😄 
I used a toast/snack to begin with with options for refreshing, I'll see if I can add it again, maybe as an option to autorefresh :+1:

---

Edit: Added toggle button to auto renew file changes. Currently user has to toggle it manually each time they open stage page. If we use it a lot let's consider adding it to persistent storage ^^

File: GitOut/Features/Git/Stage/GitStageViewModel.cs [hunk 1]
```diff
@@ -185,12 +202,87 @@ public int SelectedIndexIndex
...
+                                    );
+                                }
+                                if (selectedFileHasChanges)
+                                {
+                                    await ExecuteDiffAsync();
+                                }
+                            }
+                            else
+                            {
+                                if (!(previousCancellation is null))
```
Comment (mickedefault) [inline GitOut/Features/Git/Stage/GitStageViewModel.cs]: can be modified after the .net 5 feature

File: GitOut/Features/Git/Stage/GitStageViewModel.cs [hunk 2]
```diff
@@ -209,7 +301,9 @@ private async Task SetAppendCommitMessageAsync()
                     ? head.Subject
                     : $"{head.Subject}{Environment.NewLine}{Environment.NewLine}{head.Body}";
             }
+#pragma warning disable CA1031 // Do not catch general exception types
```
Comment (mickedefault) [inline GitOut/Features/Git/Stage/GitStageViewModel.cs:294]: this triggers a warning on my machine. Let's compare editorconfigs

File: GitOut/Features/Material/Snackbar/Snack.cs [hunk 3]
```diff
@@ -47,26 +47,28 @@ private class SnackBuilder : ISnackBuilder
             private TimeSpan? duration;
             private string? message;
             private Exception? error;
+            private CancellationToken cancelToken;
```
Comment (mickedefault) [inline GitOut/Features/Material/Snackbar/Snack.cs]: Optional:
I prefer `cancellationToken` as a name

**Review by mickedefault — CHANGES_REQUESTED** [review]
would love for a setting to always have this set!
I think we need to investigate how visual studio saves files. In my testing I didn't get the refresh button very often, because when determining if it was the current file (i.e. setting `selectedFileHasChanges`) the `args.RepositoryPath` was very often a temporary file, which I guess visual studio created. since it was a "new" file and (perhaps the files are swapped or something) gitout didn't see it as the current file.
no_refresh_button

**Comment by calahatras** [comment]
> 
> 
> would love for a setting to always have this set!
> I think we need to investigate how visual studio saves files. In my testing I didn't get the refresh button very often, because when determining if it was the current file (i.e. setting `selectedFileHasChanges`) the `args.RepositoryPath` was very often a temporary file, which I guess visual studio created. since it was a "new" file and (perhaps the files are swapped or something) gitout didn't see it as the current file.
> no_refresh_button

Interesting, I did not experience that behavior during development 🤔 If it *is* a temporary file of some kind, there's no real way for us to determine what file it actually should relate to though. It looks like the "real name" shows up with a ~-suffix, do you get multiple events for the same file? I find it strange that FileSystemWatcher doesn't get an event for the real file when that is written to in the end.

**Review by mickedefault — APPROVED** [review]

**Pull request merged**